File: src/main/java/kitchenpos/domain/product/tobe/domain/ProductName.java
package kitchenpos.domain.product.tobe.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;

import java.util.Objects;

@Embeddable
class ProductName {
    @Column(name = "name", nullable = false)
    private String name;

    protected ProductName() {
    }

    public ProductName(String name) {
        this(name, (text) -> false);
    }

    public ProductName(String name, BlackWordClient blackWordClient) {
        validateName(name);
        boolean containsProfanity = blackWordClient.containsProfanity(name);
        if (containsProfanity) {
            throw new IllegalArgumentException("name에 비속어가 포함될 수 없습니다.");
        }
        this.name = name;
    }

    private void validateName(String name) {
        if (Objects.isNull(name)) {
            throw new IllegalArgumentException("name은 Null 일 수 없습니다.");
        }
    }

    public String getName() {
        return name;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ProductName that = (ProductName) o;
        return Objects.equals(getName(), that.getName());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getName());
    }
}

File: src/main/kotlin/kitchenpos/tobe/product/domain/vo/ProductName.kt
package kitchenpos.tobe.product.domain.vo

import jakarta.persistence.Column
import jakarta.persistence.Embeddable
import kitchenpos.tobe.product.domain.ProductPurgomalumClient

@Embeddable
class ProductName private constructor(
    @Column(name = "name", nullable = false)
    val name: String,
) {
    companion object {
        fun from(
            name: String,
            productPurgomalumClient: ProductPurgomalumClient,
        ): ProductName {
            validate(name, productPurgomalumClient)
            return ProductName(name)
        }

        private fun validate(
            name: String,
            productPurgomalumClient: ProductPurgomalumClient,
        ) {
            require(name.isNotBlank()) {
                throw IllegalArgumentException("상품명은 공백이 될 수 없습니다.")
            }
            require(
                productPurgomalumClient.containsProfanity(name).not(),
            ) {
                throw IllegalArgumentException("상품명에 욕설이 포함되어 있습니다.")
            }
        }
    }
}
